test/FunctionalTests/Microsoft.OData.Core.Tests/JsonLight/ODataJsonLightCollectionSerializerTests.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                Name = "Products",
+                SerializationInfo = new ODataCollectionStartSerializationInfo
+                {
+                    CollectionTypeName = "Collection(Product)"
+                },
+                NextPageLink = new Uri("http://tempuri.org/Products?$skiptoken=Id-5"),
+                Count = 10
+            };
+
+            var itemTypeReference = new EdmEntityTypeReference(this.entityType, false);
+
+            var result = await SetupODataJsonLightCollectionSerializerAndRunTestAsync(
+                async (jsonLightCollectionSerializer) =>
+                {
+                    await jsonLightCollectionSerializer.WriteCollectionStartAsync(collectionStart, itemTypeReference);
+                    await jsonLightCollectionSerializer.WriteCollectionEndAsync();
+                },
+                /* container */ null,
+                /* writeTopLevelCollection */ true);
+
+            Assert.Equal(
+                "{\"@odata.context\":\"http://tempuri.org/$metadata#Collection(Product)\"," +
+                "\"@odata.count\":10," +
+                "\"@odata.nextLink\":\"http://tempuri.org/Products?$skiptoken=Id-5\"," +
+                "\"value\":[]}",
+                result);
+        }
+
+        private ODataJsonLightCollectionSerializer CreateODataJsonLightCollectionSerializer(bool writingResponse, IServiceProvider container = null, bool isAsync = false, bool writingTopLevelCollection = false)
+        {
+            var messageInfo = new ODataMessageInfo
+            {
+                MessageStream = this.stream,
+                MediaType = new ODataMediaType("application", "json"),
+#if NETCOREAPP1_1
+                Encoding = Encoding.GetEncoding(0),
+#else
+                Encoding = Encoding.Default,
+#endif
+                IsResponse = writingResponse,
+                IsAsync = isAsync,
+                Model = model,
+                Container = container
+            };
+            var context = new ODataJsonLightOutputContext(messageInfo, this.settings);
+            return new ODataJsonLightCollectionSerializer(context, writingTopLevelCollection);
+        }
+
+        /// <summary>
+        /// Sets up an ODataJsonLightCollectionSerializer,
+        /// then runs the given test code asynchonously,
+        /// then flushes and reads the stream back as a string for customized verification.
+        /// </summary>
+        private async Task<string> SetupODataJsonLightCollectionSerializerAndRunTestAsync(Func<ODataJsonLightCollectionSerializer, Task> func, IServiceProvider container = null, bool writingTopLevelCollection = false)
+        {
+            var jsonLightCollectionSerializer = CreateODataJsonLightCollectionSerializer(true, container, true, writingTopLevelCollection);
+            await func(jsonLightCollectionSerializer);
+            await jsonLightCollectionSerializer.JsonLightOutputContext.FlushAsync();

```

Comment:
JsonLightOutputContext.FlushAsync() will flush AsyncchronousJsonWriter?

Reply:
@xuzhg The plan is to change the current implementation of `FlushAsync` once implemented of asynchronous support is completed across all participating types. When we get there I'll modify it to flush `AsynchronousJsonWriter`